### DISCUSSION TITLE
Understanding how WebBrowserClient.SendKeyboardControls works

### DISCUSSION THREAD

**Discussion by alex-turnbull** [Voltstro-Studios/UnityWebBrowser#203] [Q&A] [answered]
As part of an AR/VR project I am creating a virtual keyboard as for interaction with the Web Browser.

```
var x = VoltstroStudios.UnityWebBrowser.Shared.WindowsKey.Enter;
VoltstroStudios.UnityWebBrowser.Shared.WindowsKey[] a = new VoltstroStudios.UnityWebBrowser.Shared.WindowsKey[1];
a[0] = x;

var y = VoltstroStudios.UnityWebBrowser.Shared.WindowsKey.None;
VoltstroStudios.UnityWebBrowser.Shared.WindowsKey[] b = new VoltstroStudios.UnityWebBrowser.Shared.WindowsKey[1];
b[0] = y;

webBrowserClient.SendKeyboardControls(a, b, new char[' ']);
```

Using this bit of crude code above I can send any respective 'command' inputs (Enter, backspace, etc) but struggle when it comes to inputting standard characters or symbols.

How would I correctly send keyboard controls to work for characters and shift modifiers?

**Comment by Voltstro** [comment, marked as answer]
If you wanted to send Ctrl+A, you could do:

```csharp
WindowsKey[] downKey = new []{ WindowsKey.A | WindowsKey.Control };
WindowsKey[] upKey = Array.Empty<WindowsKey>();
char[] chars = Array.Empty<char>();
browserClient.SendKeyboardControls(downKey, upKey, chars);
```

With capital letters you can just pass capital chars.
```csharp
WindowsKey[] downKey = Array.Empty<WindowsKey>();
WindowsKey[] upKey = Array.Empty<WindowsKey>();
char[] chars = new[] { 'A' };
browserClient.SendKeyboardControls(downKey, upKey, chars);
```

**Reply by alex-turnbull** [reply]
Thank you loads I see now.

I was doing some strange things when initialising the arrays but now I see how each parameter is used / not required for each use case.
